tiledb/api/src/column.rs

```diff
@@ -0,0 +1,416 @@
+use crate::error::Error;
+use crate::Datatype;
+use crate::Result as TileDBResult;
+
+/// The ReferenceColumn is used when we can safely use the existing Vec's or
+/// slices being used to create a column. This avoids the allocation, copy, and
+/// eventual deallocation when providing column data to TileDB.
+pub struct ReferenceColumn<'data> {
+    num_values: u64,
+    value_size: u64,
+    data: &'data [u8],
+    offsets: &'data [u64],
+}
+
+/// If a column requires allocating a linearized buffer (i.e., when passing a
+/// Vec or slice of AsRef<str> values), this structure is used to hold the
+/// allocated buffers.
+pub struct AllocatedColumn {
+    num_values: u64,
+    value_size: u64,
+    data: Box<[u8]>,
+    offsets: Box<[u64]>,
+}
+
+/// The Column struct is used to represent a column of data in TileDB. A column
+/// contains a required data slice and an optional offsets slice when the
+/// underlying column definition has a variable number of values per cell.
+///
+/// An AsColumn trait is provided for a number of common use cases such that
+/// users should be able to pass common instances directly as arguments that
+/// have the Column type.
+pub enum Column<'data> {
```

Comment:
> I'm picturing some kind of marker object which uses the `'data` lifetime to make sure that the user data lives longer than the query does. That probably looks even more different between `ReferenceColumn` and `AllocatedColumn`.

My goal with the `'data` lifetime was for the case of when we're just passing a pointer to Rust allocated RAM into TileDB so that the Column lifetime would be guaranteed to be shorter than whatever we might have borrowed from. I've got no idea if I've gotten that correct. So I should probably add a test for it specifically. And as you mentioned, those `compile_fail` tests seem like the best practice for doing that.

Reply:
Oh, I guess it is kinda bidirectional.
If you get a column back from an Enumeration then the Enumeration should out-live the column.
If you supply a column to a query (possibly where I imagine this going), then the column handle marker thing I allude to should out-live the query.

I think the use of `'data` will be able to accomplish both of these ideas